File: day5/2.js
import fs from "node:fs"
import readline from "node:readline"

// const fileStream = fs.createReadStream("input-test.txt")
const fileStream = fs.createReadStream("input.txt")
const rl = readline.createInterface({
	input: fileStream,
	crlfDelay: Infinity,
})

const ranges = []

for await (const line of rl) {
	if (line === "") break
	ranges.push(line.split("-").map(s => +s))
}

for (let i = 0; i < ranges.length; i++) {
	const range = ranges[i]
	for (let j = 0; j < ranges.length; j++) {
		if (i === j) continue
		const targetRange = ranges[j]
		if (range[0] >= targetRange[0] && range[1] <= targetRange[1]) {
			range[0] = 0
			range[1] = 0
			break
		} else if (range[0] >= targetRange[0] && range[0] <= targetRange[1]) {
			targetRange[1] = range[1]
			range[0] = 0
			range[1] = 0
			break
		} else if (range[1] <= targetRange[1] && range[1] >= targetRange[0]) {
			targetRange[0] = range[0]
			range[0] = 0
			range[1] = 0
			break
		}
	}
}

let password = 0
for (const range of ranges) {
	if (range[0] === 0 && range[1] === 0) continue
	// console.log(`-------------------------`)
	console.log(range)
	password += range[1] - range[0] + 1
}
console.log({ password })

File: day10/2.py
import pulp

def solve(buttons, joltages):
    prob = pulp.LpProblem("buttons", pulp.LpMinimize)

    # one variable per button: how many times pressed
    x = [
        pulp.LpVariable(f"x{i}", lowBound=0, cat="Integer")
        for i in range(len(buttons))
    ]

    # minimize total presses
    prob += pulp.lpSum(x)

    # exact joltage constraints
    for j in range(len(joltages)):
        prob += (
            pulp.lpSum(
                x[i] for i in range(len(buttons)) if j in buttons[i]
            )
            == joltages[j]
        )

    status = prob.solve(pulp.PULP_CBC_CMD(msg=False))
    if status != pulp.LpStatusOptimal:
        raise RuntimeError("No feasible solution")

    return int(pulp.value(prob.objective))


password = 0

# with open("input-test.txt") as f:
with open("input.txt") as f:
    for line_idx, line in enumerate(f):
        parts = line.strip().split(" ")

        # last token: joltage array
        joltages = list(map(int, parts[-1][1:-1].split(",")))

        # middle tokens: buttons
        buttons = []
        for b in parts[1:-1]:
            buttons.append(
                set(map(int, b[1:-1].split(",")))
            )

        presses = solve(buttons, joltages)
        password += presses

print(password)
